전투 로직 제작

GyutaeLee opened the pull request:
- Finite State Machine 추가
- 전투에 참여하는 객체 : Player / Enemy
- 전투에서 일어나는 액션 : 공격 / 피격
- 액션의 결과 : 체력 감소

GyutaeLee committed:
- feat : 전투 로직 제작
1. Player / Enemy 클래스 추가 및 테스트용 GamePlayer 클래스 제거
2. 테스트용 공격/피격 : player / enemy 부딪히면 서로 hp가 감소하도록 개발
3. HealthPoint 클래스 HUD hp bar 추가

GyutaeLee committed:
chore : 사용하지 않는 폴더 제거

GyutaeLee committed:
refactor : 코딩 컨벤션에 맞게 변수명 수정

GyutaeLee committed:
fix : GameLogic - GameState 딕셔너리 구조로 변경

GyutaeLee committed:
refactor : UI 클래스 네이밍 Prefix 제거

GyutaeLee committed:
refactor : Unit -> Character 네이밍 변경

GyutaeLee committed:
refactor : 프로퍼티 변수 코딩 컨벤션에 맞게 변경

GyutaeLee committed:
refactor : 누락된 Unit -> Character 네이밍 변경 추가

GyutaeLee committed:
fix : Character Health Point Head Up Dispaly 구조 변경
1. 캐릭터 부모 클래스 Character.cs 추가
2. Head Up Display 클래스를 따로 두어서 컴포넌트에 붙이면 사용할 수 있도록 수정

GyutaeLee committed:
feat : Finite State Machine 로직 추가
- reference git : https://github.com/dubit/unity-fsm

GyutaeLee committed:
feat : Game FSM 추가
1. 기존에 사용하던 FSM 로직에서 새로운 FSM 로직으로 수정 작업

GyutaeLee committed:
fix : Game FSM 로직 수정
1. GameState 내부에서 게임 관련 함수들 호출하도록 다듬음

GyutaeLee committed:
refactor : Game 관련 스크립트 네이밍 "Game" Prefix 제거

GyutaeLee committed:
feat : Enemy FSM 추가
1. 기존에 사용하던 FSM 로직에서 새로운 FSM 로직으로 수정 작업

GyutaeLee committed:
refactor : Enemy 관련 스크립트 네이밍 "Enemy" Prefix 제거

GyutaeLee committed:
refactor : Character 관련 스크립트 네이밍 "Character" Prefix 제거
1. CharacterStat -> BaseStat

GyutaeLee committed:
feat : Player FSM 추가
1. 기존에 사용하던 FSM 로직에서 새로운 FSM 로직으로 수정 작업

GyutaeLee committed:
feat : BaseStat 생성자 추가 및 멤버 변수 프로퍼티로 변경

GyutaeLee committed:
chore : namespace 형식에 맞추어 폴더 구조 변경

Parkhyunseo commented on Union/Assets/Scripts/Services/Character/Player/FiniteStateMachine.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            this._states = new List<KeyValuePair<States, State>>();
+
+            this._states.Add(new KeyValuePair<States, State>(States.Alive, new Alive(this._player, this)));
```
이런부분에서 States, State 같이 이름이 비슷할 경우 해석의 어려움을 느낄 수 있네요

GyutaeLee replied:
1. State Enum을 지우는 방향으로 시도를 해보았는데, 현재 상태를 알기가 쉽지 않을거 같아서 우선 그대로 사용하겠습니다.
2. 네이밍에 어려움을 느껴서 현재는 [StateNumber]로 변경했습니다. 회의 때도 좋은 이름이 나오지는 않았지만, 추후에 좋은 네이밍이나 방법이 생각난다면 한 번 더 바꿔도 좋을거 같습니다.

Parkhyunseo commented on Union/Assets/Scripts/Services/Character/Player/Player.cs:
```diff
@@ -0,0 +1,33 @@
+        private void Awake()
+        {
+            this.BaseStat = new BaseStat(100, 10, 10, 10, 20, 10);
+            this._finiteStateMachine = new FiniteStateMachine(this);
```
추후에 Serailzed형태로 받을 수 있게되면 좋을 것 같습니다

Parkhyunseo commented on Union/Assets/Scripts/Services/Character/Player/States.cs:
```diff
@@ -0,0 +1,76 @@
+        public abstract void Exit();
+    }
+
+    public class Alive : State
```
(질문입니다.) 상태 코드를 한 곳에 모으는게 좋을까요? 나누는게 좋을까요

GyutaeLee replied:
현재는 이대로 유지하겠습니다.

JuicyPark commented on Union/Assets/Scripts/Services/Character/Enemy/Enemy.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    public class Enemy : Character
+    {
+        private FiniteStateMachine _finiteStateMachine;
```
Enemy, Player
FiniteStateMachine의 서로 차이점이 타겟?이 Player냐 Enemey냐의 차이를 제외하면 전부 중복된 코드입니다. 하나의 FiniteStateMachine으로 합칠 수 있지않을까요?

GyutaeLee replied:
현재는 클래스의 기능과 크기가 작아서 겹치는 부분이 많습니다.
추후에 어떠한 기능이 들어갈지 모르는 상황에 합쳐두면, 추후 분리 작업에 개발 비용을 들여야할 수도 있으므로 현재는 이대로 유지하겠습니다.
(말씀하신 대로 기본 기능들이 어느 정도 정립되고 겹치는 부분이 많으면 하나의 기본 머신 클래스를 공유하고, 필요한 기능들을 추가하면 좋을거 같네요!)

Parkhyunseo commented on Union/Assets/Scripts/Services/Game/FiniteStateMachine.cs:
```diff
@@ -0,0 +1,90 @@
+
+namespace Union.Services.Game
+{
+    public class FiniteStateMachine
```
Machine과 FiniteStateMachine, Controller의 역할을 다시 한번 설명해주시겠나요

JuicyPark commented on Union/Assets/Scripts/Services/Character/Enemy/States.cs:
```diff
@@ -0,0 +1,77 @@
+﻿using UnityEngine;
```
스크립트이름이
States.cs인데
enum이름인 States를 따라가는것보단 추상클래스 State를 따라가는게 더 맞는것 같습니다

GyutaeLee replied:
수정했습니다!

JuicyPark commented on Union/Assets/Scripts/Services/Character/Enemy/FiniteStateMachine.cs:
```diff
@@ -0,0 +1,80 @@
+            }
+        }
+
+        private Enemy _enemy;
```
Enemy에서도 FiniteStateMachine를 참조하고
FiniteStateMachine에서도 Enemy를 참조하고있어서 순환참조형식인데
한곳에서 흐름이 이어지도록 개선할수 있지않을까요?

GyutaeLee replied:
하위 클래스로 순환 참조가 되지 않도록 수정했습니다.
FSM은 상태만 관리하고, 상위 클래스에서 해야할 작업은 하위 클래스인 State 클래스의 현재 상태를 확인해서 직접 작업하도록 수정했습니다.

GyutaeLee committed:
fix : Finite State Machine 수정 작업 중 (Game Logic 에만 적용 완료)
1. Machine -> FiniteStateMachine : 네이밍 변경
2. FiniteStateMachine -> FiniteStateMachineController : 변경
3. 불필요한 의존성 제거 : 각 state에 맞는 transition을 확인하고, 조건이 맞을시에 작동하도록 수정

GyutaeLee committed:
fix : Finite State Machine Controller 네이밍 변경 (-> Event Executor)

GyutaeLee committed:
fix : State Enum 네이밍 변경 ( -> StateNumber)

GyutaeLee committed:
fix : Player Finite State Machine 수정
1. 불필요한 참조 제거
2. Finit State Machine 네이밍 수정 ( -> Controller 추가)

GyutaeLee committed:
fix : Player Finite State Machine 수정
1. 불필요한 참조 제거

GyutaeLee committed:
fix : Enemy Finite State Machine 수정
1. 불필요한 참조 제거
2. Finit State Machine 네이밍 수정 ( -> Controller 추가)

GyutaeLee committed:
fix : Game FSM Controller 생성 시점에서 인스턴스 미리 준비하도록 수정

GyutaeLee committed:
chore : .gitignore 추가
